File: examples/validate_products.py
#!/usr/bin/env python3
"""
DataGuard Python Example - Product Validation

This example demonstrates how to use DataGuard to validate a CSV file
containing product data with various validation rules.

These rules match the configuration in examples/config.toml
"""

import dataguard


def main():
    print("DataGuard Example - Product Validation\n")

    # Define validation rules for each column
    # These match the rules in examples/config.toml

    # Index: integer with uniqueness, not null, and positive checks
    index_column = (
        dataguard.integer_column("Index").is_unique().is_not_null().is_positive()
    )

    # Name: string with min/max length and not null
    name_column = (
        dataguard.string_column("Name")
        .is_not_null()
        .with_min_length(3)
        .with_max_length(100)
    )

    # Description: string with min length
    description_column = dataguard.string_column("Description").with_min_length(1)

    # Currency: string with allowed values
    currency_column = dataguard.string_column("Currency").is_in(
        ["USD", "EUR", "GBP", "JPY", "CAD"]
    )

    # Price: integer with range and positive checks
    price_column = dataguard.integer_column("Price").between(1, 10000).is_positive()

    # Stock: integer with non-negative check
    stock_column = dataguard.integer_column("Stock").is_non_negative()

    # EAN: string with exact length and numeric validation
    ean_column = dataguard.string_column("EAN").is_exact_length(13).is_numeric()

    # Availability: string with enumerated values
    availability_column = dataguard.string_column("Availability").is_in(
        [
            "in_stock",
            "out_of_stock",
            "limited_stock",
            "backorder",
            "pre_order",
            "discontinued",
        ]
    )

    # Internal ID: integer with positive check
    internal_id_column = dataguard.integer_column("Internal ID").is_positive()

    # Create and configure table
    table = dataguard.CsvTable(
        "examples/data/products_large.csv", "Products Large Dataset"
    )
    table.prepare(
        [
            index_column,
            name_column,
            description_column,
            currency_column,
            price_column,
            stock_column,
            ean_column,
            availability_column,
            internal_id_column,
        ]
    )

    # Validate the data
    print("Validating products_large.csv (512k rows)...\n")
    result = table.validate()

    # Display results
    print("Validation Results:")
    print(f"  Table: {result['table_name']}")
    print(f"  Total rows: {result['total_rows']}")

    passed, total = result["passed"]
    print(f"  Rules: {passed}/{total} passed")

    if passed == total:
        print("\n✓ All validation rules passed!")
    else:
        print("\n✗ Some validation rules failed.")
        print("  Check the detailed report above for specific errors.")


if __name__ == "__main__":
    main()

File: examples/validate_products.rs
use dataguard_core::{CsvTable, NumericColumnBuilder, StringColumnBuilder, Table};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("DataGuard Example - Product Validation\n");

    // Index: integer with uniqueness, not null, and positive checks
    let mut index_column = NumericColumnBuilder::<i64>::new("Index".to_string());
    index_column
        .is_unique(0.0)
        .is_not_null(0.0)
        .is_positive(0.0);

    // Name: string with min/max length and not null
    let mut name_column = StringColumnBuilder::new("Name".to_string());
    name_column
        .is_not_null(0.0)
        .with_min_length(3, 0.0)
        .with_max_length(100, 0.0);

    // Description: string with min length
    let mut description_column = StringColumnBuilder::new("Description".to_string());
    description_column.with_min_length(1, 0.0);

    // Currency: string with allowed values
    let mut currency_column = StringColumnBuilder::new("Currency".to_string());
    currency_column.is_in(
        vec![
            "USD".to_string(),
            "EUR".to_string(),
            "GBP".to_string(),
            "JPY".to_string(),
            "CAD".to_string(),
        ],
        0.0,
    );

    // Price: integer with range and positive checks
    let mut price_column = NumericColumnBuilder::<i64>::new("Price".to_string());
    price_column.between(1, 10000, 0.0).is_positive(0.0);

    // Stock: integer with non-negative check
    let mut stock_column = NumericColumnBuilder::<i64>::new("Stock".to_string());
    stock_column.is_non_negative(0.0);

    // EAN: string with exact length and numeric validation
    let mut ean_column = StringColumnBuilder::new("EAN".to_string());
    ean_column.is_exact_length(13, 0.0);
    ean_column.is_numeric(0.0)?;

    // Availability: string with enumerated values
    let mut availability_column = StringColumnBuilder::new("Availability".to_string());
    availability_column.is_in(
        vec![
            "in_stock".to_string(),
            "out_of_stock".to_string(),
            "limited_stock".to_string(),
            "backorder".to_string(),
            "pre_order".to_string(),
            "discontinued".to_string(),
        ],
        0.0,
    );

    // Internal ID: integer with positive check
    let mut internal_id_column = NumericColumnBuilder::<i64>::new("Internal ID".to_string());
    internal_id_column.is_positive(0.0);

    let mut table = CsvTable::new(
        "examples/data/products.csv".to_string(),
        "Products Large Dataset".to_string(),
    )?;

    table.prepare(
        vec![
            Box::new(index_column),
            Box::new(name_column),
            Box::new(description_column),
            Box::new(currency_column),
            Box::new(price_column),
            Box::new(stock_column),
            Box::new(ean_column),
            Box::new(availability_column),
            Box::new(internal_id_column),
        ],
        vec![],
    )?;

    let result = table.validate()?;

    println!("Validation Results:");
    println!("  Table: {}", result.table_name);
    println!("  Total rows: {}", result.total_rows);

    let (passed, total) = result.is_passed();
    println!("  Rules: {}/{} passed", passed, total);

    if passed == total {
        println!("\n✓ All validation rules passed!");
    } else {
        println!("\n✗ Some validation rules failed.");
        println!("  Check the detailed report above for specific errors.");
    }

    Ok(())
}
